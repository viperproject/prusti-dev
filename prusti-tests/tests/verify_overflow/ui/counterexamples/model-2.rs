// compile-flags: -Punsafe_core_proof=true -Pcounterexample=true

use prusti_contracts::*;

#[trusted]
struct VecWrapper<T> {
    values: Vec<T>,
}

#[model]
struct VecWrapper<#[concrete] i32> {
    last_pushed: i32,
}

#[trusted]
#[ensures(result.model().last_pushed == val )]
fn create_vec_wrapper_i32(val: i32) -> VecWrapper<i32>{
    VecWrapper{
        values: vec![val],
    }
}


#[trusted]
#[ensures(v.model().last_pushed == val )]
fn push_i32(v: &mut VecWrapper<i32>, val: i32) {
    v.values.push(val);
}

#[requires(v.model().last_pushed == 2)] //force specific counterexample
#[ensures(v.model().last_pushed == 5)]
fn test(v: VecWrapper<i32>){}


/*
FIXME: Details: consistency error in test2: Consistency error: No matching local 
variable _1$snapshot$0 found with type Snap$ref$Unique$trusted$m_VecWrapper$I32$, 
instead found Snap$I32. (@27.11)

#[ensures(result.model().last_pushed == 5)]
fn test2(val: i32, val2: i32) -> VecWrapper<i32>{
    let mut v = create_vec_wrapper_i32(val);
    push_i32(&mut v, val2);
    v
}
*/

fn main() {}