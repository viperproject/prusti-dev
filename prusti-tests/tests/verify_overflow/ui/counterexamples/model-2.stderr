error: [Prusti: verification error] postcondition might not hold.
  --> $DIR/model-2.rs:31:11
   |
31 | #[ensures(v.model().last_pushed == 5)]
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: the error originates here
  --> $DIR/model-2.rs:32:1
   |
32 | fn test(v: VecWrapper<i32>){}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: counterexample for "v"
       value:   VecWrapper_model {
            last_pushed: 2,
        }
  --> $DIR/model-2.rs:32:9
   |
32 | fn test(v: VecWrapper<i32>){}
   |         ^
note: counterexample for "result"
       value:   ()
  --> $DIR/model-2.rs:32:28
   |
32 | fn test(v: VecWrapper<i32>){}
   |                            ^^

error: aborting due to previous error

